### PR TITLE
Linter: create rule to remove additional items when items is an object

### PR DESCRIPTION
**Pull request by Karan-Palan**
### Implementation

**Rule ID**:  dangling_additional_items
**Description**: "`additionalItems` is ignored when `items` is an object"  
**Condition**: Detects `additionalItems` in JSON Schema objects where `items` is an object (not an array).  
**Transform**: Removes the redundant `additionalItems` property.  
**Dialects**: Supports draft4, draft6, draft7, and 2019-09. Excluded for draft-2020-12 (where `additionalItems` was removed from the spec).  

  - Registered the rule in the **Common rules** section (correctness category).  
  - Added 10 test cases across all relevant JSON Schema drafts.  

### Test Cases

Positive Cases:  Removal of `additionalItems` when `items` is an object.  
Negative Cases:  No change for `items` as an array or when `items` is undefined.  
Edge Cases:  Complex `items` objects (e.g., with `if`/`then` conditions) - correctly handled.  
Context Preservation:  Tests with various other keywords to ensure proper behavior.

### CODE REVIEW

File: src/extension/alterschema/linter/dangling_additional_items.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+            const sourcemeta::core::SchemaResolver &) const
+      -> sourcemeta::core::SchemaTransformRule::Result override {
+    return contains_any(
+               vocabularies,
+               {"https://json-schema.org/draft/2019-09/vocab/applicator",
+                "http://json-schema.org/draft-07/schema#",
+                "http://json-schema.org/draft-06/schema#",
+                "http://json-schema.org/draft-04/schema#"}) &&
+           schema.is_object() && schema.defines("items") &&
+           schema.defines("additionalItems") && schema.at("items").is_object();
```
Comment (jviotti) [inline src/extension/alterschema/linter/dangling_additional_items.h:24]: Note that this applies is the schema is a booleans too! I also put comments to clarify this in the rule name and description. I think what you want to check for is if the keyword is set to a schema:

```suggestion
           schema.defines("additionalItems") && is_schema(schema.at("items"));
```

File: test/alterschema/alterschema_lint_2019_09_test.cc [hunk 2]
```diff
@@ -1661,8 +1661,7 @@ TEST(AlterSchema_lint_2019_09, unnecessary_allof_ref_wrapper_5) {
   sourcemeta::core::JSON document = sourcemeta::core::parse_json(R"JSON({
     "$schema": "https://json-schema.org/draft/2019-09/schema",
     "allOf": [
-      { 
-        "type": "integer",
+      { "type": "integer",
```
Comment (jviotti) [inline test/alterschema/alterschema_lint_2019_09_test.cc:1664]: Unintended change?

**Comment by jviotti** [comment]
I left a few comments. On testing:

- Can you test cases where `items` is set to boolean schemas?
- Can you test cases where `items` is set to an array and therefore confirm that `additionalItems` is NOT removed?

File: src/extension/alterschema/alterschema.cc [hunk 3]
```diff
@@ -52,6 +52,7 @@ static auto every_item_is_boolean(const T &container) -> bool {
 #include "linter/content_media_type_without_encoding.h"
 #include "linter/content_schema_default.h"
 #include "linter/content_schema_without_media_type.h"
+#include "linter/dangling_additional_items.h"
```
[the remark this answers is not in the document]
Comment (jviotti) [reply src/extension/alterschema/alterschema.cc:55]: Also, you have to update the `CMakeLists.txt` file to list this file too

File: src/extension/alterschema/linter/dangling_additional_items.h [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+            const sourcemeta::core::SchemaFrame::Location &,
+            const sourcemeta::core::SchemaWalker &,
+            const sourcemeta::core::SchemaResolver &) const
+      -> sourcemeta::core::SchemaTransformRule::Result override {
+    return contains_any(
+               vocabularies,
+               {"https://json-schema.org/draft/2019-09/vocab/applicator",
+                "http://json-schema.org/draft-07/schema#",
+                "http://json-schema.org/draft-06/schema#",
+                "http://json-schema.org/draft-04/schema#"}) &&
```
[the remark this answers is not in the document]
Comment (Karan-Palan) [reply src/extension/alterschema/linter/dangling_additional_items.h:22]: Yes it does, adding tests for it as well

**Comment by Karan-Palan** [comment]
- Renamed the rule, class, file, and all references.
- Updated CMakeLists.txt
- Added Draft 3 support with tests.
- Added comprehensive tests for boolean items, array items, object items, and scenarios where the rule should _not_ apply.

@jviotti , the PR is ready for review

File: src/extension/alterschema/linter/additional_items_with_schema_items.h [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+      -> sourcemeta::core::SchemaTransformRule::Result override {
+    return contains_any(
+               vocabularies,
+               {"https://json-schema.org/draft/2019-09/vocab/applicator",
+                "http://json-schema.org/draft-07/schema#",
+                "http://json-schema.org/draft-06/schema#",
+                "http://json-schema.org/draft-04/schema#",
+                "http://json-schema.org/draft-03/schema#"}) &&
+           schema.is_object() && schema.defines("items") &&
+           schema.defines("additionalItems") && schema.at("items").is_object();
```
Comment (jviotti) [inline src/extension/alterschema/linter/additional_items_with_schema_items.h:25]:
```suggestion
           schema.defines("additionalItems") && is_schema(schema.at("items"));
```

[the remark this answers is not in the document]
Comment (jviotti) [reply src/extension/alterschema/linter/additional_items_with_schema_items.h:25]: Also I think you are missing test cases covering that case: `items` set to a boolean

**Comment by Karan-Palan** [comment]
@jviotti , resolved merge conflicts. Any comments?
@relequestual, any other tests that come into your mind?
